### ISSUE TITLE
Improve performance

### ISSUE THREAD

**Issue by deeprobin** [deeprobin/bruteforce-rs#1]
The function `raw_next` builds the entire string from scratch every time it is called when it could be reusing a buffer and just changing the last character each time.

Maybe with `String.pop()` and `String.push(c)`

**Comment by deeprobin** [comment]
It is also possible to return `&str` instead of `String`. It would return a reference to the content of the String so that there are zero copies.
However, it won't be possible to `impl Iterator` -> "Streaming Iterator problem"

**Comment by deeprobin** [comment]
Pull request #2 optimized performance of test `test_cracker` from 2,4s to 2,0s

**Comment by deeprobin** [comment]
Current bench results: 
```
robin@ROBIN-PC MINGW64 /f/bruteforce (master)
$ cargo bench
   Compiling bruteforce v0.1.3 (F:\bruteforce)
    Finished bench [optimized] target(s) in 2.20s
     Running target\release\deps\bruteforce-5846197cc1289a50.exe

running 4 tests
test tests::bench_are_all_chars_last ... bench:          11 ns/iter (+/- 2)
test tests::bench_new                ... bench:         174 ns/iter (+/- 27)
test tests::bench_next               ... bench:       1,039 ns/iter (+/- 174)
test tests::bench_next_char          ... bench:           2 ns/iter (+/- 0)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured; 0 filtered out
```

**Comment by chinoto** [comment]
I have an idea that should make this much more efficient and reduce the code quite a bit. For a charset of A-Z, you could think of BruteForce.current as similar to a number with a radix of 26 that you use addition with carry over on, so when you add one to "AZ" [0,25] you get "BA" [1,0] and for "ZZ" [25,25] you get "AAA" [0,0,0].
I considered a radix of 27 where 0 mean nothing, but [1,0] doesn't make sense ("A?"), so instead carry over on the last digit means pushing a 0 to the vector rather than a 1. Since pushing to a vector is more efficient at the end rather than the beginning, the end of the string would be the beginning of the vector. Generating a string from the vector would simply be `current_vec.iter().rev().map(|n| self.charset[n]).collect::<String>();`.

**Comment by deeprobin** [comment]
@chinoto May be a good idea.

**Comment by deeprobin** [comment]
New bench results: 
```
test tests::bench_are_all_chars_last ... bench:          10 ns/iter (+/- 4)
test tests::bench_new                ... bench:         163 ns/iter (+/- 19)
test tests::bench_next               ... bench:         925 ns/iter (+/- 185)
test tests::bench_next_char          ... bench:           2 ns/iter (+/- 0)
```

We should really optimize raw_next/next.

@chinoto had a good idea
> 
> 
> I have an idea that should make this much more efficient and reduce the code quite a bit. For a charset of A-Z, you could think of BruteForce.current as similar to a number with a radix of 26 that you use addition with carry over on, so when you add one to "AZ" [0,25] you get "BA" [1,0] and for "ZZ" [25,25] you get "AAA" [0,0,0].
> I considered a radix of 27 where 0 mean nothing, but [1,0] doesn't make sense ("A?"), so instead carry over on the last digit means pushing a 0 to the vector rather than a 1. Since pushing to a vector is more efficient at the end rather than the beginning, the end of the string would be the beginning of the vector. Generating a string from the vector would simply be `current_vec.iter().rev().map(|n| self.charset[n]).collect::<String>();`.

**Comment by deeprobin** [comment]
raw_next builds also the String from scratch instead of changing just the changed characters. Maybe it could be also useful to use &str instead of String because it is a primitive with a pointer.

**Comment by deeprobin** [comment]
In rust `char` is a 4-byte integer. It is better to use &str as charset.

**Comment by chinoto** [comment]
A char array may be larger than the corresponding string, but I'm betting the char array is faster to access individual chars since there isn't any parsing needed. Bench it and see, maybe I'm wrong.

**Comment by deeprobin** [comment]
New bench results after commit https://github.com/DeepRobin/bruteforce-rs/commit/4b36b7df3be6a844bbff225a0fd42a89f89b353f 

```
running 7 tests
test tests::test_combined_charset ... ignored
test tests::test_multibyte_char ... ignored
test tests::bench_are_all_chars_last ... bench:          10 ns/iter (+/- 1)
test tests::bench_new                ... bench:         180 ns/iter (+/- 67)
test tests::bench_next               ... bench:         884 ns/iter (+/- 66)
test tests::bench_next_char          ... bench:           2 ns/iter (+/- 0)
test tests::bench_raw_next           ... bench:         733 ns/iter (+/- 137)```

**Comment by chinoto** [comment]
Using pointers should be slower due to the indirection and on a 64 bit OS, a pointer is 8 bytes, whereas a char is only 4. The change makes no noticeable difference on my computer.

**Comment by chinoto** [comment]
I managed to get some benefits from parallelization in an experiment, now I just have to figure out how to carry the idea over into this project.
https://github.com/chinoto/string_finder

**Comment by chinoto** [comment]
What do you think of pulling in the arrayvec crate so that ArrayString can be used? It would significantly reduce the allocations needed. In my test, switching to it resulted in a time of 250ms for both single threaded and multi threaded, where before the switch I got 800ms for single threaded and 6500ms for multi threaded (overhead from allocator mutex [more info](https://www.embedded.com/tackling-memory-allocation-in-multicore-and-multithreaded-applications/)). 

I've found a more efficient way of generating all character set combinations for increasing lengths in my experimentation repo. It's rather long due to trying to optimize for Rayon (no benefit or loss...), but pretty short and simple when implemented for just a single thread.
https://github.com/chinoto/string_finder/blob/master/src/bin/for_flat_map.rs

**Comment by deeprobin** [comment]
Yes that could be a good solution, but i have this week no time to implement that in this repo.

**Comment by chinoto** [comment]
Just remembered the two `new_*()` functions. I can easily have this version start at a specified length, but starting at a specified string could be complex unless I make the first call to `raw_next()` take awhile by just using `.skip_while()` (every iteration up to the given string has to be computed and discarded).

**Comment by deeprobin** [comment]
Huh, that's not so good, because maybe people want to save the current state and "resume" it at next program start

**Comment by deeprobin** [comment]
Here's an idea:
You implement a feature for rayon that allows you to use a parallel iterator. But you don't have to, because rayon is optional as a feature.

**Comment by chinoto** [comment]
I am already using rayon in string_finder, no additional features are needed in rayon to make it work.  I could probably get the skip to happen on each character until the target string is reached.

While I found this to be a neat project at first, I probably won't do anything major with it until you make it actually do something useful. I thought I had said this before, but apparently that was specifically in response to making an OpenCL implementation in the ["Add GPU calculation" issue](https://github.com/deeprobin/bruteforce-rs/issues/10#issuecomment-579674079).

**Comment by deeprobin** [comment]
@chinoto I tried to use [smallvec](https://docs.rs/smallvec/1.2.0/smallvec/) instead of Vec, but I saw it is slower than the std-Vec.

My current implementation: [➤ branch: smallvec-usage](https://github.com/deeprobin/bruteforce-rs/tree/smallvec-usage)

```
test tests::bench_charset_by_range   ... bench:       1,094 ns/iter (+/- 1,032)
test tests::bench_charset_concat     ... bench:       2,555 ns/iter (+/- 1,839)
test tests::bench_charset_new        ... bench:           1 ns/iter (+/- 3)
test tests::bench_charset_new_by_str ... bench:       1,367 ns/iter (+/- 1,067)
test tests::bench_charset_to_string  ... bench:      11,482 ns/iter (+/- 12,788)
test tests::bench_new                ... bench:          11 ns/iter (+/- 5)
test tests::bench_next               ... bench:         191 ns/iter (+/- 158)
test tests::bench_raw_next           ... bench:          44 ns/iter (+/- 32)```

**Comment by deeprobin** [comment]
We now use [criterion](https://crates.io/crates/criterion) for the benchmarks (see pull request #23)

**Comment by deeprobin** [comment]
The performance was improved again (pull request #25).

**Comment by deeprobin** [comment]
And again (pull request #24)

**Comment by chinoto** [comment]
I rebased smallvec-usage onto master (https://github.com/deeprobin/bruteforce-rs/commit/b2c1a317c0976910dc2e14eb3624b37996439b1e) and changed the size to 32 so that it actually has a chance to be on the stack, but there is a 5.9% increase in time over master.

After thinking about it for awhile, the reason is that while your raw_current is on the heap, it hardly ever has to reallocate. On the other hand, my implementation was creating a new Vec on the heap every iteration, so changing it to the stack was a significant help in my case.

BruteForce::next() creates a new string on the heap on every call, so it could benefit from being on the stack if you don't mind returning a non-standard type to the library user. Here are a couple crates for that.
[inlinable_string](https://crates.io/crates/inlinable_string): Been around for three years; has 12 releases; has no hard dependencies.
[smallstr](https://crates.io/crates/smallstr): Been around for 2 years; has 2 releases; basically a wrapper over smallvec, which means it benefits from all the work done on smallvec and has less of its own code that isn't tested due to lack of popularity.
